gh-113212: Improve error message & document zero-arg super inside nested functions and generator expressions (#113307)
python/cpython: Doc/library/functions.rst

Reviewer: Ideally, I think the code examples here would take up less vertical space on the page. This PR adds quite a few lines; it would be nice to make it more concise

Author: Reused example from above to make it more concise

   the second argument is a type, ``issubclass(type2, type)`` must be true (this
   is useful for classmethods).

>    When called directly within an ordinary method of a class, both arguments may
>    be omitted ("zero-argument :func:`super`"). In this case, *type* will be the
>    enclosing class, and *obj* will be the first argument of the immediately
>    enclosing function (typically ``self``). (This means that zero-argument
>    :func:`super` will not work as expected within a nested function or a generator
>    expression, which implicitly creates a nested function.)

Reviewer: Some very minor suggested wording tweaks:
- No need to link to the function reference we're already in; use `!` to prevent Sphinx from adding links when `super` is referenced
- Some suggested changes to the last sentence to make it more elegant.

```suggestion
   When called directly within an ordinary method of a class, both arguments may
   be omitted ("zero-argument :func:`!super`"). In this case, *type* will be the
   enclosing class, and *obj* will be the first argument of the immediately
   enclosing function (typically ``self``). (This means that zero-argument
   :func:`!super` will not work as expected within nested functions, including
   generator expressions, which implicitly create nested functions.)
```

Author: 👍

              super().method(arg)    # This does the same thing as:
                                     # super(C, self).method(arg)

>    .. warning::

Reviewer: `.. warning` blocks are rendered very dramatically with our current theme, such that I think I'd probably only use them for known security vulnerabilities:

<details>
<summary>Screenshot</summary>

![20E3DAC1-E8FB-4640-A48B-9CF6D6AA2E9F](https://github.com/python/cpython/assets/66076021/e2598ac8-23f0-4fea-88b5-d6267b96a3f1)

</details>

I'd suggest using `.. note` instead.
```suggestion
   .. note::
```

In my experience, this kind of thing comes up relatively rarely, so I'd probably also put the note at the bottom of the docs for super(), rather than in the middle

Author: 👍

>       By default, if no second argument is provided, :func:`super` uses first argument

Reviewer: ```suggestion
      By default, if no second argument is provided, :func:`super` uses the first argument
```

Author: 👍

>       of immediately enclosing function definition which is ``self`` under normal circumstances.

Reviewer: ```suggestion
      of the immediately enclosing function definition, which is ``self`` under normal circumstances.
```

Author: 👍

      This may lead to unexpected behaviour if used inside nested functions or generator expressions

>       as the latter creates implicit nested function. Following example will result in exception::

Reviewer: ```suggestion
      (the latter creates implicit nested functions). The following example will result in an exception::
```

Author: 👍

   :func:`super`, see `guide to using super()
   <https://rhettinger.wordpress.com/2011/05/26/super-considered-super/>`_.

   .. note::
      By default, if no second argument is provided, :func:`super` uses the first argument
      of the immediately enclosing function definition, which is ``self`` under normal circumstances.
      This may lead to unexpected behaviour if used inside nested functions or generator expressions
      (the latter creates implicit nested functions). The following example will result in an exception::

         class B(C):
            def method(self, arg):
               list(super().method(arg) for _ in range(1))   # Raises TypeError.

      In such cases, you should provide arguments explicitly::

         class B(C):
            def method(self, arg):

>                list(super(B, self).method(arg) for _ in range(1))

Reviewer 2: I think the clarification of what zero-argument super is and how it works should a) be shorter and more streamlined than this, and b) belongs higher up in the flow. One reason the current doc is awkward is that the "typical use" example shows zero-arg super even though the prior text never mentioned or explained it at all; the current PR doesn't fix that awkwardness.

I also don't think the generator expression case is important enough to deserve so much vertical space and its own entire pair of extra code examples. A simple mention is plenty.

Here is how I would propose to modify the docs: https://gist.github.com/carljm/098a8ed9ab387f244c1e15a16bc193db

Author: First version was right after first code example but @AlexWaygood suggested this belonged to the bottom. Also, why is it problematic to have a code example? It is always a lot better to have a visual example that goes along with textual explanation IMHO.

Reviewer: > First version was right after first code example but @AlexWaygood suggested this belonged to the bottom.

If you go with a shorter version that is more of a description of the semantics rather than a warning about a specific footgun (like @carljm suggests in https://gist.github.com/carljm/098a8ed9ab387f244c1e15a16bc193db), then I think it will be fine to have it in the middle of the docs :)

> It is always a lot better to have a visual example that goes along with textual explanation IMHO.

Examples can definitely be helpful, but we do users a disservice if we present them with a wall of text that we expect them to wade through. It's very valuable to keep docs concise for things like this, and trying to user `super()` in a generator expression is something of an edge case

Reviewer 2: @WolframAlph I agree that in general examples make a concept easier to understand. But there are many different concepts involved in `super()`, and a large array of possible code examples we could show to illustrate different usages. We have a limited "budget" for users' attention in the docs, and we have to prioritize. It doesn't make sense to me that two out of three total code examples in the `super()` documentation would be dedicated to illustrating the very specific edge case of usage in generator expressions. It's better to focus their attention on the single already-existing example demonstrating the most common and best usage, rather than dedicating code examples to showing what will not work.

Author: 👍 , updated PR
